### DISCUSSION TITLE
Help with custom builds for Nodemcu (having esp8266ex).

### DISCUSSION THREAD

**Discussion by gammerboy** [arendst/Tasmota#12475] [Q&A] [answered]
Will adding the below line for a new env in platformio_tasmota_cenv.ini and building compile the correct bin for nodemcu. Do I have to change any other parameters.
board                       = esp8266_4M2M

eg.
[env:tasmota-tas]
board                       = esp8266_4M2M
build_flags = ${common.build_flags} -DFIRMWARE_TAS

Do i have to Include the below line along with board:
;board_build.f_cpu       = 160000000L
;board_build.f_flash     = 40000000L

Also went through #10528 merge request and noticed that about 1M is available for OTA does that mean I can flash a bin greater than 500kb <1mb ?

Did try the above without flashing a minimal bin but for some reason seems to slow down the web server. Had to reflash using tasmotizer which fixes the issue. Tried later by flashing a minimal file built with the 4m2m board parameter added in cenv.ini file and then the new bin where I didn't face the slowdown.

Another question. I am testing the VL53L0X sensor in tasmota for measuring water level. Will changing parameters in xsns_45_vl53l0x.ino under the tasmota folder be carried into each new build I make in VS or do I have to clear some kind of a previous build cache to get the new settings in. If so how should I be doing it.

Thank you.

**Comment by barbudor** [comment, marked as answer]
I'm not sure what your suggest change will do
Normally what is supported is to change the linker file to enable flash above 1MB for file system. See `platfoirmio_override_sample.ini`:
```
; Build variant 1MB = 1MB firmware no filesystem (default)
;board_build.ldscript        = eagle.flash.1m.ld
; Build variant 2MB = 1MB firmware, +744k OTA, 256k filesystem (Zigbee Bridge, most Shelly devices)
;board_build.ldscript        = eagle.flash.2m256.ld
; Build variant 4MB = 1MB firmware, +1MB OTA, 2MB filesystem (WEMOS D1 Mini, NodeMCU, Sonoff POW)
;board_build.ldscript        = eagle.flash.4m2m.ld
; Build variant 16MB = 1MB firmware, +1MB OTA, 14MB filesystem (WEMOS D1 Mini pro, Ledunia (=32MB))
;board_build.ldscript        = eagle.flash.16m14m.ld
```

With default map, the total sizeof(current_uncompressed_firmware) + sizeof(minimal_firmware) should fit into the flash (around 950KB)
With others, obvioulsy you will have room for larger OTA

**Reply by gammerboy** [reply]
Hey sorry the current version of the override sample file has 

```
; Build variant 1MB = 1MB firmware no filesystem (default)
board                   = ${common.board}
; Build variant 2MB = 1MB firmware, 1MB filesystem (most Shelly devices)
;board                   = esp8266_2M1M
; Build variant 4MB = 1MB firmware, 1MB OTA, 2MB filesystem (WEMOS D1 Mini, NodeMCU, Sonoff POW)
;board                   = esp8266_4M2M
;board_build.f_cpu       = 160000000L
;board_build.f_flash     = 40000000L
```

As for flash size. What is the expected behavior when you try flashing a bin greater than available on flash? Does the upload just fail? 

As for the first question do I have to add any parameters to the buildparameters or are there some other options that I might have to change?

Here is some of the first few relevant lines from build output:

PLATFORM: Espressif 8266 (2.6.3) > Espressif Generic ESP8266 Tasmota 1M sketch 1M OTA 2M SPIFFS
HARDWARE: ESP8266 80MHz, 80KB RAM, 972.00KB Flash

I can see that it is going for 4mb kinda but the second line says 972.00kb flash so am I doing something wrong or is this expected. 

Just trying to figure out if I've done things correctly and also as to why flashing ota update without flashing minimal is causing slowdown.

Thank you.

**Reply by barbudor** [reply]
Sorry, you are right this was changed a few days ago. I didn't noticed it
So `;board                   = esp8266_4M2M` shoud be the way to go
I'm currently updating my build env to match and testing it currently

Flashing a bin greater than flash size is probably a bad idea but I can't tell if this will be detected or not.
Also it is not recommended to change from a map mode to another through OTA. Erase and serial-reflash is recommended
Although some cases may work

I will invite @Jason2866 to comment further

**Comment by Jason2866** [comment]
Just take a look in the file `platformio_tasmota_env.ini` how a valid [custom-env] has to look like.
You only have to add the settings which differs from the [env] setting (first setting in the file)
Example
```
[env:tasmota-mybuild]
board                   = esp8266_4M2M
build_flags             = ${env.build_flags} -DFIRMWARE_MYBUILD
```
Attention `common.build_flags` has changed to `env.build_flags` in latest development version!

**Reply by gammerboy** [reply]
Hey .. Thank you .. Have changed to env.build_flags.

**Comment by Jason2866** [comment]
In the boards file the maxsize of Tasmota code is defined. So building a Tasmota binary which is overboarded in size should generate a warning and/or a linking error.
Every! ESP8266 programm can use 1M max. for code. All other flash space can only be used for OTA or as filesystem.
And yes with 1M OTA space you can flash every build without minimal or using .gz

**Reply by gammerboy** [reply]
Can the compressed .gz file for lets say a 520kb bin file be used to update using ota without using minimal ? Does it not take space to unpack the file in some way or does it just  start replacing the files in "active space" (I'm not sure what it might actually be called) kinda

**Reply by Jason2866** [reply]
Yes. Thats how .gz works. Avoids minimal in many cases.
